### DISCUSSION TITLE
Object recognition without AVX

### DISCUSSION THREAD

**Discussion by el-fredo** [immich-app/immich#987] [General]
I know that Immich needs a CPU with AVX support to use the machine learning features. Due to the huge increase in energy costs, I decided to go with a very energy efficient home server, which is powered by an Intel Pentium Silver J5040. Unfortunately, this CPU does not support AVX.

Then I found this thread: [Machine learning on CPUs without AVX: howto](https://github.com/immich-app/immich/discussions/300) 
It explains how to use the whole TensorFlow thing on CPUs without AVX. Unfortunately, this tutorial is for professionals, because you have to create containers and build / compile TensorFlow (as far as I understand).

Do you think there might be a simpler solution in the future? Could this method perhaps be built directly into Immich? So I mean that the object detection in Immich is working with every CPU?

I would also offer myself as a test subject if someone could take me by the hand. Then I would put together a step-by-step guide for beginners. However, I would also like to mention that I have Immich running under Unraid. Maybe that makes things even more complicated?

**Comment by bertmelis** [comment]
I wrote the "tutorial" and I am by far not a professional. Besides, further down that thread you can find a precompiled tfjs. I created it on a J4105.

Where do you need help? I can guide you through the process and improve the documentation while I do this.

**Comment by el-fredo** [comment]
Thank you so much for offering your help. Unfortunately, I have no idea about Docker and Docker-Compose. From that you are a pro for me ;-) In Unraid I am only using the community appstore to install containers.

For an absolute beginner like me, it would probably be easier to use your precompiled images. I think our CPUs are almost identical. So as the first step, I download your libtensorflow.tar.gz from your Google Drive. Then I need to install this as a container in Unraid. I do that with Docker Compose, right?

My first question is, where to store the downloaded file and do I have to unpack it?

Second question: Should I use the docker-compose.yml I already used to install Immich on Unraid, and just modify the part:
```
immich-machine-learning:
    image: altran1502/immich-machine-learning:release
```
Or should I create a new docker-compose.yml, just for this new machine learning image?

**Reply by bertmelis** [reply]
Well, I don't know a thing about Unraid. Can you ash into your server? Or do you have a terminal on your server GUI?

**Reply by el-fredo** [reply]
Yes, Unraid provides a terminal. I can also open a terminal in each installed container.

**Comment by el-fredo** [comment]
I have come one step further:

Using the command `docker import /mnt/user/download/libtensorflow.tar.gz immich-ml-new:latest` in the Unraid terminal, I was able to import your image.

So when I type `docker image ls`, I see all installed images, and also the new `immich-ml-new:latest` (that's the name I have chosen).

The next step now is to create the Docker container, right? And this is where I get stuck. Can you tell me what the Docker command should be?

For simplicity, I tried to create the container using Portainer. There I edited the existing container `immich-immich-machine-learning-1`, but with your image instead of the original one. Unfortunately, I get the error 400.

**Reply by bertmelis** [reply]
Wait a minute. The tar is not a docker image but the result of tfjs compilation.

The way you use it is to copy it into the official immich-ml image. By doing this you overwrite the tfjs that doesn't work.

I don't have access to my pc now so bit difficult to support ATM.

**Reply by el-fredo** [reply]
Okay, then I still haven't understood the whole principle, unfortunately. I'm also stuck because I am missing not only the Docker knowledge, but also how Unraid handles Docker. Images, building containers etc... Don't know where to start.

Would it not be possible to provide the complete "new" container? So that this can be pulled by interested people (instead of the official one)? My first post was actually aimed at this, namely to find a simple solution for people who don't have an AVX-compatible CPU.

**Reply by bertmelis** [reply]
Maybe I can provide an onofficial non-avx container. I will look into this but this for me is uncharted territory.

**Reply by bo0tzz** [reply]
@bertmelis if tensorflow is built for non-avx, will that build then work for all the non-avx machines or is it closely tied to the machine it was built on? If the former, we could maybe look into supplying an 'official' non-avx image.

**Reply by bertmelis** [reply]
I really don't know. I've now built tfjs and specified Bazel to build for "native" so it might not be universal. But I'm sure it is possible to create a universal build. It's all a new experience for me and learn while doing.

I also don't know if it can be automated on GitHub. I vaguely remember there are restrictions on duration of the runners.

**Reply by bo0tzz** [reply]
I imagine (but could be mistaken) that the runner that makes that build needs to not have AVX itself? If so, we would need to supply our own runner anyways, so duration restrictions would not be an issue.

**Reply by bertmelis** [reply]
I'm talking about compiling tfjs. It just takes a long time and I don't know if it's possible to build it on GitHub. Disabling avx is done by a compiler flag so the machine that does the build doesn't matter (although I'm even less familiar with cross-compilation).

**Reply by bertmelis** [reply]
Some points:
- A github action can run up to 6 hours. This might be enough to build Tensorflow.js from source. Maybe a self-hosted runner might also be needed.
- Tensorflow.js is built with the default `-march=native` flag which implies `-mtune=native`. So by default it is optimized for use with the machine where it is built. this is easily changed to `-mtune=generic` to fit most cpus at the expense of optimizations.

**Reply by bo0tzz** [reply]
The current ML container is mainly a placeholder and is essentially never updated. That means that build times aren't really an issue since we can cache approximately everything. On the other hand, it raises the question of whether it's worth putting effort into this at all, since we'll probably be throwing it out in the future.

**Reply by el-fredo** [reply]
Hi @bo0tzz, what exactly do you mean by

> since we'll probably be throwing it out in the future

Do you mean Immich will lose the machine learning feature?

**Reply by bo0tzz** [reply]
No worries, I just mean in the future we'll replace it with a better one.

**Reply by el-fredo** [reply]
Okay, great.

I had briefly tested Photoprism, here the object detection worked perfectly (with my CPU without AVX support). But I prefer Immich, because it is exactly what I have been searching for.

The background why I startet this thread: In Unraid I have to pull a total of 7 containers to get Immich running. And I just thought it would be easy to replace the original ML-container with one that does not require AVX. So users can choose which one they pull, depending on their CPU.

In any case, thanks for your efforts 👍🏼

**Reply by bertmelis** [reply]
I'll still create the non-avx version but I'll build tensorflow manually. I only have different priorities now. Have a few days?

**Reply by bertmelis** [reply]
@el-fredo Would this be easier?
https://github.com/bertmelis/immich-machine-learning-no-avx

**Reply by el-fredo** [reply]
It is so much easier! Thank you so much for your effort.

All I had to do was change the single line in my docker-compose.yml and click on "Update Stack". That was everything I had to do. It pulled your container, which took a while. I restarted Immich, went to Administration -> Jobs and clicked on "Detect Objects" -> Run. Now my server is busy. Very busy... More than 24,000 jobs to do, the CPU load of all four cores is at 99 - 100% 😏

In the smartphone app, I already see the "Things" section when I tap Search. A few objects are already listed there, so it seems to work.

Thank you very much again!

**Reply by bertmelis** [reply]
It's indeed quite a large image because it add another layer on top of an already large image. It's quick and dirty but it works.

**Comment by alextran1502** [comment]
@bertmelis Thank you for the support! I appreciate your continuous responses to this topic.

**Reply by bertmelis** [reply]
With pleasure. I now published my very first image. Learning by doing.
